controllers/space/space_controller.go

```diff
@@ -0,0 +1,343 @@
+//+kubebuilder:rbac:groups=toolchain.dev.openshift.com,resources=spaces/finalizers,verbs=update
+
+// Reconcile ensures that there is an NSTemplateSet resource defined in the target member cluster
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx, "namespace", r.Namespace)
+	logger.Info("reconciling Space")
+
+	// Fetch the Space
+	space := &toolchainv1alpha1.Space{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{
+		Namespace: r.Namespace,
+		Name:      request.Name,
+	}, space)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("Space not found")
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, errs.Wrap(err, "unable to get the current Space")
+	}
+	if !util.IsBeingDeleted(space) {
+		// Add the finalizer if it is not present
+		if err := r.addFinalizer(logger, space); err != nil {
+			return reconcile.Result{}, err
+		}
+	} else {
+		return reconcile.Result{}, r.ensureSpaceDeletion(logger, space)
+	}
+	// ensure that there's a NSTemplateSet on the Target Cluster
+	// will trigger a requeue until the NSTemplateSet exists and is ready,
+	// so the Space can be in `ready` status as well
+	requeue, err := r.ensureNSTemplateSet(logger, space)
+	return ctrl.Result{Requeue: requeue}, err
+}
+
+// setFinalizers sets the finalizers for Space
+func (r *Reconciler) addFinalizer(logger logr.Logger, space *toolchainv1alpha1.Space) error {
+	// Add the finalizer if it is not present
+	if !util.HasFinalizer(space, toolchainv1alpha1.FinalizerName) {
+		logger.Info("adding finalizer on Space")
+		util.AddFinalizer(space, toolchainv1alpha1.FinalizerName)
+		if err := r.Client.Update(context.TODO(), space); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ensureNSTemplateSet creates the NSTemplateSet on the target member cluster if it does not exist,
+// and updates the space's status accordingly.
+// returns `true` after creating the NSTemplateSet and until it is `ready`
+func (r *Reconciler) ensureNSTemplateSet(logger logr.Logger, space *toolchainv1alpha1.Space) (bool, error) {
+	if space.Spec.TargetCluster == "" {
+		return false, r.setStatusProvisioningFailed(logger, space, fmt.Errorf("unspecified target member cluster"))
+	}
+	memberCluster, found := r.MemberClusters[space.Spec.TargetCluster]
+	if !found {
+		return false, r.setStatusProvisioningFailed(logger, space, fmt.Errorf("unknown target member cluster '%s'", space.Spec.TargetCluster))
+	}
+	name := space.Name // NSTemplateSet will have the same name as the Space
+	logger = logger.WithValues("target_member_cluster", space.Spec.TargetCluster, "name", name)
```

Comment:
correct me if I'm wrong, but the name should have been included in the logger as part of the logger initialization at the very beginning of the reconcile.

Reply:
removed in 6415808